iotdb/src/main/java/org/apache/iotdb/db/sync/sender/FileSenderImpl.java

```diff
@@ -0,0 +1,507 @@
+      LOGGER.error("Can not make fileSnapshot");
+      throw new IOException(e);
+    }
+    return validFilesSnapshot;
+  }
+
+  /**
+   * Transfer data of a storage group to receiver.
+   *
+   * @param fileSnapshotList list of sending snapshot files in a storage group.
+   */
+  public void syncData(Set<String> fileSnapshotList) throws SyncConnectionException {
+    try {
+      int successNum = 0;
+      for (String snapshotFilePath : fileSnapshotList) {
+        successNum++;
+        File file = new File(snapshotFilePath);
+        List<String> filePathSplit = new ArrayList<>();
+        String os = System.getProperty("os.name");
+        if (os.toLowerCase().startsWith("windows")) {
+          String[] name = snapshotFilePath.split(File.separator + File.separator);
+          filePathSplit.add(name[name.length - 2]);
+          filePathSplit.add(name[name.length - 1]);
+        } else {
+          String[] name = snapshotFilePath.split(File.separator);
+          filePathSplit.add(name[name.length - 2]);
+          filePathSplit.add(name[name.length - 1]);
+        }
+        while (true) {
+          // Sync all data to receiver
+          byte[] buffer = new byte[Constans.DATA_CHUNK_SIZE];
+          int data;
+          try (FileInputStream fis = new FileInputStream(file)) {
+            ByteArrayOutputStream bos = new ByteArrayOutputStream(Constans.DATA_CHUNK_SIZE);
+            while ((data = fis.read(buffer)) != -1) { // cut the file into pieces to send
+              bos.write(buffer, 0, data);
+              ByteBuffer buffToSend = ByteBuffer.wrap(bos.toByteArray());
+              bos.reset();
+              serviceClient
+                  .syncData(null, filePathSplit, buffToSend, SyncDataStatus.PROCESSING_STATUS);
+            }
+            bos.close();
+          }
+
+          // Get md5 of the file.
+          MessageDigest md = MessageDigest.getInstance("MD5");
+          try (FileInputStream fis = new FileInputStream(file)) {
+            while ((data = fis.read(buffer)) != -1) {
+              md.update(buffer, 0, data);
+            }
+          }
+
+          // the file is sent successfully
+          String md5OfSender = (new BigInteger(1, md.digest())).toString(16);
+          String md5OfReceiver = serviceClient.syncData(md5OfSender, filePathSplit,
+              null, SyncDataStatus.SUCCESS_STATUS);
+          if (md5OfSender.equals(md5OfReceiver)) {
+            LOGGER.info("Receiver has received {} successfully.", snapshotFilePath);
+            break;
+          }
+        }
+        if (LOGGER.isInfoEnabled()) {
+          LOGGER.info(String.format("Task of synchronization has completed %d/%d.", successNum,
+              fileSnapshotList.size()));
+        }
+      }
+    } catch (Exception e) {
+      throw new SyncConnectionException("Cannot sync data with receiver.", e);
+    }
+  }
+
+  /**
+   * Sync schema with receiver.
+   */
+  @Override
+  public void syncSchema() throws SyncConnectionException {
```

Comment:
The schema file does not need to be verified by MD5?

Reply:
Actually, it's better to verify by MD5, fixed